### PR TITLE
Handle the `const struct *` and `struct *` patterns

### PR DESCRIPTION
**Pull request by pvdrz**
Given that C keeps a different namespace for `struct`s and aliases. The following pattern
```c
typedef const struct foo {
    void *bar;
} *foo;
```
is valid C code and produces both a `struct` and a pointer called `foo`. Given that Rust does not make this distinction, we add the `_ptr` prefix to the pointer type alias to avoid any name collisions.

Fixes: https://github.com/rust-lang/rust-bindgen/issues/2227

### CODE REVIEW

File: bindgen/ir/ty.rs [hunk 1]
```diff
@@ -1099,6 +1100,20 @@ impl Type {
                         // within the clang parsing.
                         TypeKind::Opaque
                     } else {
+                        // Check if this type definition is an alias to a pointer of a `const
+                        // struct` with the same name and add the `_ptr` suffix to it to avoid name
+                        // collisions.
+                        if !ctx.options().c_naming {
+                            if let Some(inner_name) = dbg!(inner_spelling
+                                .strip_prefix("const struct ")
```
Comment (goffrie) [inline bindgen/ir/ty.rs:1108]: This seems not specific to pointers-to-`const` - wouldn't something like `typedef struct foo { ... } *foo;` have the same issue?

File: bindgen/ir/ty.rs [hunk 2]
```diff
@@ -1099,6 +1100,20 @@ impl Type {
...
+                                .and_then(|s| s.strip_suffix(" *")))
```
Comment (goffrie) [inline bindgen/ir/ty.rs:1109]: This textual approach seems unreliable; what if there isn't a space before the `*`? Possibly other cases?

Comment (pvdrz) [reply bindgen/ir/ty.rs:1109]: This textual representation is generated by `clang` and it doesn't seem to be affected by adding/removing spaces from the source code. For example:
```
typedef const struct foo {
    char bar;
}*foo;
```
works just fine.

**Comment by pvdrz** [comment]
@goffrie thanks for your review! I updated the PR with some changes if you want to give it a second look

File: bindgen/ir/ty.rs [hunk 3]
```diff
@@ -1099,6 +1100,23 @@ impl Type {
                         // within the clang parsing.
                         TypeKind::Opaque
                     } else {
+                        // Check if this type definition is an alias to a pointer of a `const
+                        // struct` with the same name and add the `_ptr` suffix to it to avoid name
+                        // collisions.
+                        if !ctx.options().c_naming {
+                            if let Some(inner_name) = inner_spelling
+                                .strip_prefix("const struct ")
```
Comment (emilio) [inline bindgen/ir/ty.rs:1108]: This feels a bit hackish, relying on libclang's serialization... Can we use `pointee_type()` and cursor equality instead? If not I guess this is ok...

Comment (pvdrz) [reply bindgen/ir/ty.rs]: I did a commit using `pointee_type` but I don't get the cursor equality part, wdym?

**Review by emilio — CHANGES_REQUESTED** [review]

**Review by emilio — APPROVED** [review]
Thanks!

**Pull request merged**